awspub/notification.py

```diff
@@ -0,0 +1,94 @@
+"""
+Methods used to handle notifications for AWS using SNS
+"""
+
+import json
+import logging
+from typing import Any, Dict
+
+import boto3
+from mypy_boto3_sns.client import SNSClient
+
+from awspub.exceptions import AWSNotificationException
+
+logger = logging.getLogger(__name__)
+
+
+class SNSNotification(object):
+    """
+    A data object that contains validation logic and
+    structuring rules for SNS notification JSON
+    """
+
+    def __init__(self, conf: Dict[str, Any], image_id: str, region_name: str):
+        """
+        Construct a message and verify that it is valid
+        """
+        self._conf = conf
+        self._image_id = image_id
+        self._region_name = region_name
+
+    def _get_topic(self, client: SNSClient) -> str:
+        """
+        Get topic ARN
+
+        :param client: AWS boto3 client with SNS service
+        :type client: SNSClient
+        :return: return topic ARN or empty string if topic does not exist
+        :rtype: str
+        """
+        resp = client.list_topics()
+        for topic in resp["Topics"]:
+            # Matching topic name at the end of ARN
+            if self.topic_name in topic["TopicArn"]:
+                return topic["TopicArn"]
+        return ""
+
+    def _get_subject(self) -> str:
+        """
+        Get subject line for config
+
+        :return: string of subject line
+        :rtype: str
+        """
+
+        return self.topic_config["subject"]
+
+    def _get_message(self) -> str:
+        """
+        Get message dictionary from config and
+        convert to JSON string type
+
+        :return: string block of email template
+        :rtype: str
+        """
+
+        return json.dumps(self.topic_config["message"])
+
+    def publish(self) -> None:
+        """
+        send notification to subscribers
+        """
+
+        snsclient: SNSClient = boto3.client("sns", region_name=self._region_name)
+        # Create topic if not exist
+        for topic in self._conf["sns"]:
+            self.topic_name = next(iter(topic))
+            self.topic_config = topic[self.topic_name]
+
+            logger.info("Checking if topic exists")
+            if not self._get_topic(snsclient):
+                logger.error(f"Topic {self.topic_name} is not exist. Unable to send SNS notification.")
+            else:
+                # Publish email
+                logger.info(f"The Topic {self.topic_name} exists. Sending an SNS notification to users...")
+                try:
+                    snsclient.publish(
+                        TopicArn=self._get_topic(snsclient),
+                        Subject=self._get_subject(),
+                        Message=self._get_message(),
+                        MessageStructure="json",
+                    )
+                except Exception as e:
+                    raise AWSNotificationException(str(e))
+                logger.info(f"The SNS notification titled '{self._get_subject()}' has been sent successfully.")
```

Comment:
region and topic name are useful info here, too. So
`Sent SNS message {subject} to topic {topic} in region {self._region_name}`

Reply:
Fixed message